kura/test/org.eclipse.kura.core.configuration.test/src/test/java/org/eclipse/kura/core/configuration/ConfigurationServiceTest.java

```diff
@@ -0,0 +1,806 @@
+import org.eclipse.kura.core.configuration.metatype.Tocd;
+import org.eclipse.kura.core.testutil.TestUtil;
+import org.eclipse.kura.crypto.CryptoService;
+import org.junit.FixMethodOrder;
+import org.junit.Test;
+import org.junit.runners.MethodSorters;
+import org.mockito.Mockito;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
+import org.osgi.service.cm.Configuration;
+import org.osgi.service.cm.ConfigurationAdmin;
+
+@FixMethodOrder(MethodSorters.NAME_ASCENDING)
+public class ConfigurationServiceTest {
+
+    @Test
+    public void testGetFactoryComponentPids() throws NoSuchFieldException {
+        // test that the returned PIDs are the same as in the service and that they cannot be modified
+
+        String[] expectedPIDs = { "pid1", "pid2", "pid3" };
+
+        ConfigurationService cs = new ConfigurationServiceImpl();
+
+        @SuppressWarnings("unchecked")
+        Set<String> s = (Set<String>) TestUtil.getFieldValue(cs, "m_factoryPids");
+        s.addAll(Arrays.asList(expectedPIDs));
+
+        Set<String> factoryComponentPids = cs.getFactoryComponentPids();
+
+        assertEquals("same length", 3, factoryComponentPids.size());
+
+        Object[] pidsArray = factoryComponentPids.toArray();
+        Arrays.sort(pidsArray);
+
+        for (int i = 0; i < pidsArray.length; i++) {
+            assertEquals(expectedPIDs[i], expectedPIDs[i], pidsArray[i]);
+        }
+
+        try {
+            factoryComponentPids.add("unsupported");
+            fail("Updating PIDs should not be possible.");
+        } catch (UnsupportedOperationException e) {
+            // OK
+        }
+    }
+
+    @Test
+    public void testCreateFactoryConfigurationNulls() {
+        // negative test; how null values are handled
+
+        ConfigurationService cs = new ConfigurationServiceImpl();
+
+        String factoryPid = null;
+        String pid = null;
+        Map<String, Object> properties = null;
+        boolean takeSnapshot = false;
+
+        try {
+            cs.createFactoryConfiguration(factoryPid, pid, properties, takeSnapshot);
+
+            fail("Exception expected with null pid value.");
+        } catch (Exception e) {
+            // OK, probably
+        }
+    }
+
+    @Test
+    public void testCreateFactoryExistingPid() throws KuraException, IOException, NoSuchFieldException {
+        // negative test; what if existing PID is used
+
+        final String factoryPid = "fpid";
+        final String pid = "mypid";
+        Map<String, Object> properties = null;
+        final boolean takeSnapshot = false;
+
+        ConfigurationServiceImpl cs = new ConfigurationServiceImpl();
+
+        Map<String, String> pids = (Map<String, String>) TestUtil.getFieldValue(cs, "m_servicePidByPid");
+        pids.put(pid, pid);
+
+        try {
+            cs.createFactoryConfiguration(factoryPid, pid, properties, takeSnapshot);
+
+            fail("Exception expected with existing pid value.");
```

Comment:
Why aren't you declaring an expected Exception on the unit test itself my @Test(expected=?)?